### PR TITLE
View context

### PR DESCRIPTION
**Pull request by LoganWright**
#1036

- Needs a test or two.
- Potentially needs discussion about deprecation solution and if there are alternatives that might be preferable.

### CODE REVIEW

File: Sources/Vapor/View/ViewData.swift [hunk 1]
```diff
@@ -0,0 +1,64 @@
+/// Structured Data Wrapper for view specific operations
+public struct ViewData: StructuredDataWrapper {
+    public static let defaultContext = ViewContext.shared
```
Comment (tanner0101) [inline Sources/Vapor/View/ViewData.swift]: does this need `Context?`

File: Sources/Vapor/View/ViewData.swift [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+public protocol ViewDataRepresentable {
+    func makeViewData() throws -> ViewData
+}
+
+public protocol ViewDataInitializable {
+    init(viewData: ViewData) throws
+}
+
+public protocol ViewDataConvertible: ViewDataRepresentable, ViewDataInitializable {}
```
Comment (tanner0101) [inline Sources/Vapor/View/ViewData.swift]: should we start moving to composition early where we can?

Comment (tanner0101) [inline Sources/Vapor/View/ViewData.swift]: referencing https://github.com/vapor/node/issues/80

File: Sources/Vapor/View/ViewData.swift [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    public init(viewData: ViewData) {
+        self = viewData
+    }
+}
+
+extension ViewData: FuzzyConverter {
+    public static func represent<T>(
+        _ any: T,
+        in context: Context
+        ) throws -> Node? {
```
Comment (tanner0101) [inline Sources/Vapor/View/ViewData.swift]: spacing

File: Sources/Vapor/View/ViewData.swift [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        guard let r = any as? ViewDataRepresentable else {
+            return nil
+        }
+
+        return try r.makeViewData().converted()
+    }
+
+    public static func initialize<T>(
+        node: Node
+        ) throws -> T? {
```
Comment (tanner0101) [inline Sources/Vapor/View/ViewData.swift]: mark

File: Sources/Vapor/View/ViewRenderer.swift [hunk 5]
```diff
@@ -7,27 +7,82 @@
...
+
+    /// Creates a view at the supplied path
+    /// using a Node that is made optional
+    /// by various protocol extensions.
+    func make(_ path: String, _ data: ViewData) throws -> View
+
     /// Creates a view at the supplied path
     /// using a Node that is made optional
     /// by various protocol extensions.
     func make(_ path: String, _ context: Node) throws -> View
```
Comment (tanner0101) [inline Sources/Vapor/View/ViewRenderer.swift]: adding a `FIXME: Deprecated` here will remind us to remove in Vapor 3

File: Sources/Vapor/View/ViewRenderer.swift [hunk 6]
```diff
@@ -7,27 +7,82 @@
...
+        let viewsDir = provider?.viewsDir ?? ""
+        return try make(viewsDir + path, viewData)
+
+    }
+
+    public func make(
+        _ path: String,
+        from provider: Provider.Type? = nil,
+        _ data: () throws -> ViewData
+        ) throws -> View {
```
Comment (tanner0101) [inline Sources/Vapor/View/ViewRenderer.swift]: mark

File: Sources/Vapor/View/ViewData.swift (hunk 1 above)
Comment (LoganWright) [inline Sources/Vapor/View/ViewData.swift]: oh shit, yes, yes it does

**Pull request merged**